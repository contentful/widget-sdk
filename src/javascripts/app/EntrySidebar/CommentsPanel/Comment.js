import React, { useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { css, cx } from 'emotion';
import {
  Heading,
  Tooltip,
  Notification,
  ModalConfirm,
  Paragraph,
  IconButton
} from '@contentful/forma-36-react-components';
import tokens from '@contentful/forma-36-tokens';
import { remove as removeComment } from 'data/CMA/CommentsRepo';
import { canRemoveComment } from './utils';
import * as types from './CommentPropTypes';

export const styles = {
  comment: css({
    display: 'flex',
    flexDirection: 'column',
    marginBottom: tokens.spacingM,
    '&:last-child': {
      marginBottom: 0
    }
  }),
  header: css({
    display: 'flex',
    alignItems: 'flex-start',
    marginBottom: tokens.spacingM
  }),
  avatar: css({
    width: 36,
    height: 36,
    background: tokens.colorElementLight,
    borderRadius: '100%',
    marginRight: tokens.spacingS
  }),
  avatarSmall: css({
    width: 18,
    height: 18
  }),
  meta: css({
    display: 'flex',
    justifyContent: 'space-between',
    flexGrow: 1,
    flexDirection: 'column'
  }),
  userName: css({
    fontSize: tokens.fontSizeM
  }),
  timestamp: css({
    color: tokens.colorTextLight,
    fontSize: tokens.fontSizeS
  }),
  commentBody: css({
    wordBreak: 'break-word',
    whiteSpace: 'pre-wrap'
  }),
  commentBodyIndented: css({
    marginLeft: `calc(${tokens.spacingS} + 18px)` // For aligning comment body with avatar
  })
};

// TODO: Make this a dumb component like <Task /> neither triggering any CMA requests
//  directly, nor receiving an entire CMA `Comment` object.

export default function Comment({ endpoint, comment, onRemoved, className, hasReplies, isReply }) {
  const {
    sys: { createdBy, createdAt }
  } = comment;
  const creationDate = moment(createdAt, moment.ISO_8601);
  const handleRemove = async () => {
    const entry = comment.sys.reference || comment.sys.parentEntity;
    try {
      await removeComment(endpoint, entry.sys.id, comment);
      onRemoved(comment);
    } catch (err) {
      Notification.error(err.message);
    }
  };

  return (
    <div className={cx([styles.comment, className])} data-test-id="comment">
      <header className={styles.header}>
        <img
          className={cx(styles.avatar, isReply && styles.avatarSmall)}
          src={createdBy.avatarUrl}
          data-test-id="comment.avatar"
        />
        <div className={styles.meta}>
          <Heading element="h4" className={styles.userName} data-test-id="comment.user">
            {renderUserName(createdBy)}
          </Heading>
          <time
            dateTime={creationDate.toISOString()}
            title={creationDate.format('LLLL')}
            className={styles.timestamp}
            data-test-id="comment.timestamp">
            {creationDate.fromNow()}
          </time>
        </div>
        <CommentActions comment={comment} onRemove={handleRemove} hasReplies={hasReplies} />
      </header>
      <div
        className={cx(styles.commentBody, isReply && styles.commentBodyIndented)}
        data-test-id="comment.body">
        {comment.body}
      </div>
    </div>
  );
}

Comment.propTypes = {
  endpoint: PropTypes.func.isRequired,
  comment: types.Comment.isRequired,
  onRemoved: PropTypes.func.isRequired,
  className: PropTypes.string,
  // Only used to check if the comment can be removed.
  // Get rid of this once thread deletion is supported by the API
  hasReplies: PropTypes.bool,
  isReply: PropTypes.bool
};

function renderUserName(user) {
  return user.firstName ? (
    <React.Fragment>{`${user.firstName} ${user.lastName}`}</React.Fragment>
  ) : (
    <Tooltip content="The author of this comment is no longer a member of this organization">
      {'(Deactivated user)'}
    </Tooltip>
  );
}

function CommentActions({ comment, onRemove, hasReplies }) {
  const [showRemovalDialog, setShowRemovalDialog] = useState(false);

  if (hasReplies || !canRemoveComment(comment)) return null;

  return (
    <React.Fragment>
      <IconButton
        iconProps={{
          icon: 'Delete'
        }}
        buttonType="muted"
        testId="comment.menu.remove"
        label="Delete comment"
        onClick={() => setShowRemovalDialog(true)}
      />
      <RemovalConfirmationDialog
        isShown={showRemovalDialog}
        onConfirm={() => {
          setShowRemovalDialog(false);
          onRemove();
        }}
        onCancel={() => setShowRemovalDialog(false)}
        data-test-id="comment.removal-confirmation"
      />
    </React.Fragment>
  );
}
CommentActions.propTypes = {
  comment: types.Comment.isRequired,
  onRemove: PropTypes.func.isRequired,
  // Only used to check if the comment can be removed.
  // Get rid of this once thread deletion is supported by the API
  hasReplies: PropTypes.bool
};

function RemovalConfirmationDialog({ isShown, onConfirm, onCancel }) {
  return (
    <ModalConfirm
      isShown={isShown}
      title="Remove comment"
      intent="negative"
      onCancel={onCancel}
      onConfirm={onConfirm}>
      <Paragraph>Are you sure you want to remove this comment?</Paragraph>
    </ModalConfirm>
  );
}
RemovalConfirmationDialog.propTypes = {
  isShown: PropTypes.bool.isRequired,
  onConfirm: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired
};
