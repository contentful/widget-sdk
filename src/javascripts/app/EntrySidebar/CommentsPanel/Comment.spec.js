import React from 'react';
import { shallow } from 'enzyme';
import Comment from './Comment.es6';

const unknownAuthor = { sys: { id: 'abc' } };
const author = { firstName: 'John', lastName: 'Doe', avatarUrl: '0.jpeg', sys: { id: 'abc' } };
const date = '2019-01-01T10:00:00.000Z';
const withAuthor = {
  body: 'Foo',
  sys: {
    createdAt: new Date(date).toISOString(),
    createdBy: author,
    id: 'xyz'
  }
};
const withoutAuthor = {
  body: 'Boo',
  sys: {
    createdAt: new Date(date).toISOString(),
    createdBy: unknownAuthor,
    id: 'xyz'
  }
};

describe('Comment', () => {
  const render = comment => shallow(<Comment comment={comment} />);
  const now = new Date(date).valueOf();

  beforeEach(() => {
    jest.spyOn(Date, 'now').mockImplementation(() => now);
  });

  afterAll(() => {
    Date.now.mockRestore();
  });

  it('renders a comment with the author info', () => {
    const component = render(withAuthor);
    expect(component).toMatchSnapshot();
  });

  it('renders a comment without the author info', () => {
    const component = render(withoutAuthor);
    expect(component).toMatchSnapshot();
  });
});
