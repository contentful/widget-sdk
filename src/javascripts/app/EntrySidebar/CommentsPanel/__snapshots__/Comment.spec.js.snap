// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment renders a comment with the author info 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
}

.emotion-6:last-child {
  margin-bottom: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  margin-bottom: 1rem;
}

.emotion-0 {
  width: 36px;
  height: 36px;
  background: #e5ebed;
  border-radius: 100%;
  margin-right: 0.75rem;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  font-size: 0.875rem;
}

.emotion-2 {
  color: #a9b9c0;
  font-size: 0.75rem;
}

.emotion-5 {
  word-break: break-word;
  white-space: pre-wrap;
}

<div
  className="emotion-6"
>
  <header
    className="emotion-4"
  >
    <img
      className="emotion-0"
      src="0.jpeg"
    />
    <div
      className="emotion-3"
    >
      <Heading
        className="emotion-1"
        element="h4"
        testId="cf-ui-heading"
      >
        John Doe
      </Heading>
      <time
        className="emotion-2"
        dateTime="2019-01-01T10:00:00.000Z"
        title="Tuesday, January 1, 2019 10:00 AM"
      >
        a few seconds ago
      </time>
    </div>
    <CommentActions
      comment={
        Object {
          "body": "Foo",
          "sys": Object {
            "createdAt": "2019-01-01T10:00:00.000Z",
            "createdBy": Object {
              "avatarUrl": "0.jpeg",
              "firstName": "John",
              "lastName": "Doe",
              "sys": Object {
                "id": "abc",
              },
            },
            "id": "xyz",
          },
        }
      }
      onRemove={[Function]}
    />
  </header>
  <div
    className="emotion-5"
  >
    Foo
  </div>
</div>
`;

exports[`Comment renders a comment without the author info 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
}

.emotion-6:last-child {
  margin-bottom: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  margin-bottom: 1rem;
}

.emotion-0 {
  width: 36px;
  height: 36px;
  background: #e5ebed;
  border-radius: 100%;
  margin-right: 0.75rem;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  font-size: 0.875rem;
}

.emotion-2 {
  color: #a9b9c0;
  font-size: 0.75rem;
}

.emotion-5 {
  word-break: break-word;
  white-space: pre-wrap;
}

<div
  className="emotion-6"
>
  <header
    className="emotion-4"
  >
    <img
      className="emotion-0"
    />
    <div
      className="emotion-3"
    >
      <Heading
        className="emotion-1"
        element="h4"
        testId="cf-ui-heading"
      >
        <Tooltip
          containerElement="span"
          content="The author of this comment is no longer a member of this organization"
          isVisible={false}
          maxWidth={360}
          place="top"
          testId="cf-ui-tooltip"
        >
          (Deactivated user)
        </Tooltip>
      </Heading>
      <time
        className="emotion-2"
        dateTime="2019-01-01T10:00:00.000Z"
        title="Tuesday, January 1, 2019 10:00 AM"
      >
        a few seconds ago
      </time>
    </div>
    <CommentActions
      comment={
        Object {
          "body": "Boo",
          "sys": Object {
            "createdAt": "2019-01-01T10:00:00.000Z",
            "createdBy": Object {
              "sys": Object {
                "id": "abc",
              },
            },
            "id": "xyz",
          },
        }
      }
      onRemove={[Function]}
    />
  </header>
  <div
    className="emotion-5"
  >
    Boo
  </div>
</div>
`;
